### PR TITLE
gyp: fix XCode CLT version detection on Catalina

### PR DESCRIPTION
**Pull request by cclauss**
Copies @szekelyisz nodejs/node-gyp#2141 to this repo.

### CODE REVIEW

**Comment by rvagg** [comment]
@szekelyisz: any idea what `/usr/sbin/softwareupdate --history` does with software that's since been removed? Is it possible that this will find an old CLI tools that is no longer there?

**Comment by szekelyisz** [comment]
> @szekelyisz: any idea what `/usr/sbin/softwareupdate --history` does with software that's since been removed? Is it possible that this will find an old CLI tools that is no longer there?

Hmm I didn't test that. AFAIK there's no (clean) way to roll back any update in macOS, and the package is gone, so the only way I can imagine to remove the tools is to delete the installation directory (usually `/Library/Developer/CommandLineTools/`). But in that case this patch will do exactly the same as happened before (compilation will fail with `command not found` or something).

File: pylib/gyp/xcode_emulation.py [hunk 1]
```diff
@@ -1541,6 +1541,13 @@ def CLTVersion():
         except GypError:
             continue
 
+    regex = re.compile(r'Command Line Tools for Xcode\s+(?P<version>\S+)')
+    try:
+        output = GetStdout(['/usr/sbin/softwareupdate', '--history'])
+        return re.search(regex, output).groupdict()['version']
```
Comment (targos) [inline pylib/gyp/xcode_emulation.py:1547]:
```suggestion
        output = GetStdout(["/usr/sbin/softwareupdate", "--history"])
        return re.search(regex, output).groupdict()["version"]
```

I think `black` would add double quotes here. Could we lint for it?

**Comment by BethGriggs** [comment]
As @richardlau mentioned, I had seemingly related issue with Xcode on Catalina yesterday. 

Before: 
```
creating config.gypi
creating config.status
creating config.mk
running: 
    python tools/gyp_node.py --no-parallel -f make-mac
No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.
No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.
No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.
gyp: No Xcode or CLT version detected!
Error running GYP
```

After applying the patch from https://github.com/nodejs/node-gyp/pull/2141 I was able to build (this was v10.x). The `No receipt for ...` warnings still appeared in the output though. I ended up just `sudo rm -r /Library/Developer/CommandLineTools` and reinstalling.

**Comment by lorenzopicoli** [comment]
When can we expect to have this merged and released?

**Comment by ryzokuken** [comment]
I'll merge this now since @BethGriggs has confirmed that it fixes the problem atleast to some extent. At some point in the near future, either me or @targos would make a release and update it in nodejs/node and nodejs/node-gyp.

**Pull request merged**